File: January/Day12/leetcode_1519.cpp
class Solution {
public:
    void dfs(string&labels, vector<vector<int>>&adj, vector<int>&ans,vector<int>&mp,int curr,int prev)
    {
       int prevCount = mp[labels[curr]-'a'];
       mp[labels[curr]-'a']++;

       for(int&ele : adj[curr])
       {
          if(ele!= prev)
          {
              dfs(labels,adj,ans,mp,ele,curr);
          } 
       }
       ans[curr] = mp[labels[curr]-'a']-prevCount;
    }
    vector<int> countSubTrees(int n, vector<vector<int>>& edges, string labels) {
      
      vector<vector<int>>adj(n);
      for(auto ele : edges)
      {
        adj[ele[0]].push_back(ele[1]);
        adj[ele[1]].push_back(ele[0]);
      }
      vector<int>ans(n);
      vector<int>mp(26,0);
      dfs(labels,adj,ans,mp,0,-1);
      return ans;

    }
};

File: January/Day12/leetcode_1519.java
class Solution {
    
    public int[] countSubTrees(int n, int[][] edges, String labels) {
        int []ans=new int[n];
        List<List<Integer>> ls = new ArrayList<>();
        for(int i=0;i<n;i++){
            ls.add(new ArrayList<Integer>(0));
        }
        for(int i=0;i<edges.length;i++){
            ls.get(edges[i][0]).add(edges[i][1]);
            ls.get(edges[i][1]).add(edges[i][0]);
        }
        boolean [] visited=new boolean[n];
        int [] label= new int[26];
         dfs(ls,labels,visited,label,0,ans);
         return ans;
        
    }
    private void dfs(List<List<Integer>> ls,String labels,boolean [] visited,int [] label,int i,int []ans){
        if(visited[i]==true)return;
        visited[i]=true;
        int prevCount = label[labels.charAt(i)-'a'];
        label[labels.charAt(i)-'a']++;
        for(int index:ls.get(i)){
            if(visited[index]!=true){
                dfs(ls,labels,visited,label,index,ans);
               
            }


        }
        ans[i]=label[labels.charAt(i)-'a']-prevCount;
         //label=new int[26];
        return ;
    }
}
